### PR TITLE
Test: Write tests for input validations on the Register form

### LINKED ISSUE #28: Test: Write tests for input validations on the Register form
## Description
The Register component has many input fields. And it raises or displays a validation message if any of the inputs are not according to the format expected. 

## Acceptance Criteria
- [ ] Write tests using React Testing Library in the Register.test.js file 
- [ ] Test if validation message shows up for name 
- [ ] Test if validation message shows up for username
- [ ] Test if validation message shows up for email
- [ ] Test if validation message shows up for password

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 2 mentor.

## Estimation
3 hours

### PR DESCRIPTION
**Pull request by Aaishpra**
### Description

Working on the following

- [x] Write tests using React Testing Library in the Register.test.js file

- [x] Test if validation message shows up for name

- [x] Test if validation message shows up for username

- [x] Test if validation message shows up for email

- [x] Test if validation message shows up for password

- [x] Test for empty field 

Fixes #28 

### Type of Change:

- Code
- Quality Assurance

**Code/Quality Assurance Only**
- Bug fix (non-breaking change which fixes an issue)
- This change requires a documentation update (software upgrade on readme file)
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### Checklist:

Will 
- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings
- [x] My PR currently breaks something (fix or feature that would cause existing functionality to not work as expected)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been published in downstream modules

### CODE REVIEW

**Comment by Aaishpra** [comment]
@mtreacy002 the empty field test is passing can You help me with the validation message test.
Another question should I use toBeValid or ToHaveContext for Validation  message test

File: src/register/Register.jsx
Comment (mtreacy002) [inline src/register/Register.jsx:90]: Add aria-labelledby attribute like how it is done in the `PersonalDetails.jsx`
Screen Shot 2020-09-03 at 7 03 37 pm

as it is pointed out by the warning
Screen Shot 2020-09-03 at 7 08 07 pm

Do this to all other input fields where `required` attribute is used.

File: src/tests/Register.test.js
Comment (mtreacy002) [inline src/tests/Register.test.js]: Be careful of indentation. This is a new test so it doesn't need to be indented.

File: src/tests/Register.test.js
Comment (mtreacy002) [inline src/tests/Register.test.js]: Remember to remove unused line.

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
@Aaishpra , you're heading the right direction. Just a little bit to go, I believe. Check out my comments below.

**Comment by mtreacy002** [comment]
> @mtreacy002 the empty field test is passing can You help me with the validation message test.
> Another question should I use toBeValid or ToHaveContext for Validation message test

@Aaishpra , you can do with whichever one works for you. 😉

**Comment by Aaishpra** [comment]
@mtreacy002 while running the second test i am getting this error
image
and while doing getAllByLabelText i am getting the following error
image
and thought of using toHavecontext

File: src/register/Register.jsx
Comment (mtreacy002) [inline src/register/Register.jsx]: the convention is to use camel-case (start with lower case) on naming attributes like `id, aria-labelledby`, so use `name` instead of `Name`

Comment (mtreacy002) [reply src/register/Register.jsx]: please apply this in other relevant places.

File: src/tests/Register.test.js
Comment (mtreacy002) [inline src/tests/Register.test.js]: please use the present tense with lower case so it reads well as a test name (`it checks empty field warning` instead of `it Check empty field warrning`)

File: src/tests/Register.test.js
Comment (mtreacy002) [inline src/tests/Register.test.js]: You can either declare as cons at the beginning to be used in the test or don't declare at all and use`screen.getByLabelText...` all the way. But to be consistent with the first test (it allows user to register succcessfully) keep it as `screen.getByLabelText`.  Keeping consistency in how the code is written across the project helps improve code readability.

File: src/tests/Register.test.js
Comment (mtreacy002) [inline src/tests/Register.test.js]: screen.getByLabelText...

File: src/tests/Register.test.js
Comment (mtreacy002) [inline src/tests/Register.test.js]: Same here, don't declare constant.

Plus, these 2 lines (105 -106) actually causing the error with muliple text called "Name" is detected since the page being rendered twice (you can see that reading through the error message with the html structures)

File: src/register/Register.jsx
Comment (mtreacy002) [inline src/register/Register.jsx]: id not Id (lower case i). This is pointed out by the error message. Check other place in the code as well.

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
@Aaishpra , please check my feedback below. You're almost there. 😉

File: src/register/Register.jsx
Comment (mtreacy002) [inline src/register/Register.jsx]: please use camel-case here (and everywhere else), so "errorName" instead of "errorname"

**Review by mtreacy002 — COMMENTED** [review]
Adding to the review

**Comment by Aaishpra** [comment]
@mtreacy002 getting the following error now for the second test
image
but the email test seems to work with a warning
image

**Comment by Aaishpra** [comment]
@mtreacy002 can you look into this and tell me where i am going wrong

File: src/tests/Register.test.js
Comment (mtreacy002) [inline src/tests/Register.test.js]: I don't see you have the test IDs for elements you're referring to, which is why the error is there. You could instead use the `.getByLabelText` by making use of the `aria-labelledby` attributes that are already in place within the validation message elements. 
also, try to be consistent with the element you're targetting on the await waitForElement() (for e.g line 130 above).
See an example of how .getByLabelText() is used on line 64 and 66)
Apply this to the rest of validation errors as well.

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
@Aaishpra. Please check my comment below.

**Comment by Aaishpra** [comment]
@mtreacy002 used getByLabelText and every test works well except the password one, getting the following error, 
image
but in my view, I have used the aria-labelled by attribute correctly as it is working completely fine with other fields. Please look into this

**Review by mtreacy002 — COMMENTED** [review]
@Aaishpra , can you fix the merge conflict first? Then I can pull your latest change and look for what causing the error.

**Comment by Aaishpra** [comment]
@mtreacy002 are the conflicts resolved now? If not can I open a new pull request and make the changes there and close this one

**Comment by mtreacy002** [comment]
> @mtreacy002 are the conflicts resolved now? If not can I open a new pull request and make the changes there and close this one

@Aaishpra . No need to open a new PR just because this one is having conflict. After contributing for a while, you'll notice that merge conflicts will occur every now and then. This is a normal thing since we're working with many other contributors. 
I'll check it out and let you know.

**Comment by mtreacy002** [comment]
@Aaishpra . I've checked your conflict reolution and found that you've removed a test put in placed by other contributor that got merged into develop. Can you put this back?
Screen Shot 2020-09-07 at 2 44 35 pm

**Comment by Aaishpra** [comment]
@mtreacy002 should i put it before my tests or after them

**Comment by mtreacy002** [comment]
> @mtreacy002 should i put it before my tests or after them

The order won't matter, @Aaishpra. each test works separately from the others.

**Comment by Aaishpra** [comment]
@mtreacy002 should I just copy and paste already merged test while editing the file and commit it, or something else is needed. Actually this is the first time that I got merge conflicts so don't have much idea of this.

**Comment by mtreacy002** [comment]
@Aaishpra, for now you can just copy paste from the already merged test. However, it's also good to learn the proper way on solving merge conflict. There are many resources you can find on how to resolve merge conflicts. Github has their own documentation [here](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/resolving-a-merge-conflict-on-github). I personally like to use `git pull --rebase`. [Here](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/resolving-a-merge-conflict-on-github) is just an example of a discussion on stackoverflow on git pull --rebase. I also wrote a [blog](https://medium.com/anitab-org-open-source/how-i-managed-to-not-mess-up-my-git-history-thanks-to-git-pull-rebase-fed452c661f0?source=friends_link&sk=e1f05ca4aec8265d26a5f05e2be545f0) about that during my gsoc journey. 
Let me knoe if you have any question regarding this.

**Comment by Aaishpra** [comment]
@mtreacy002 please notify me if the latest change is not fine I will surely use git -rebase command then, for now, I have just copied and pasted it as you mentioned, and thanks for your consideration.

File: src/tests/Register.test.js
Comment (mtreacy002) [inline src/tests/Register.test.js:148]: Add another fireEveent.change for `Confirm Password :` field here to prevent error

File: src/tests/Register.test.js
Comment (mtreacy002) [inline src/tests/Register.test.js:162]: Same here, need to add an assert for `Confirm Password :`

File: src/tests/Register.test.js
Comment (mtreacy002) [inline src/tests/Register.test.js:185]: same as above, add a fireEvent for `Confirm Password :`

File: src/tests/Register.test.js
Comment (mtreacy002) [inline src/tests/Register.test.js]: From line 172-175, you are targetting the error message inside the spans, not the input fields. So, use aria-label attribute on each span like how it was used on responseMessage in Register.jsx line 248.

Screen Shot 2020-09-08 at 6 35 38 pm

And you can use it inside test likee how it was used in first test `allows the user to register successfully` line 69.
Screen Shot 2020-09-08 at 6 39 21 pm

File: src/tests/Register.test.js
Comment (mtreacy002) [inline src/tests/Register.test.js:200]: Don't forget to add another assert for Confirm Password error message.

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
Hi @Aaishpra . You're very close now. Please check my feedback.

File: src/tests/Register.test.js
Comment (Aaishpra) [reply src/tests/Register.test.js]: @mtreacy002 can I use the following code for validation message warning all the tests are passing with this code
image
cuz I did every change as u requested me to but, the password test is still failing, the other one seems to work fine.

Comment (mtreacy002) [reply src/tests/Register.test.js]: If it works for you, then go ahead with it 👍

Comment (mtreacy002) [reply src/tests/Register.test.js]: Don't forget to push the latest change.

Comment (Aaishpra) [reply src/tests/Register.test.js]: > Don't forget to push the latest change.

Sure

**Comment by Aaishpra** [comment]
@mtreacy002 all the tests are passing just getting this warning
image

**Review by mtreacy002 — APPROVED** [review]
Congrats, @Aaishpra for passing all the tests. I don't think we need to worry about the warning message for now. We just need one more reviewer to approve this PR before it can be merged. cc @anitab-org/bridgeintech-maintainers and @anitab-org/coding-team 

Thank you so much for your patience and contribution to the project.

**Comment by Aaishpra** [comment]
Thanks alot @mtreacy002 for your guidance.❤️

**Comment by mtreacy002** [comment]
> Thanks alot @mtreacy002 for your guidance.❤️

You are most welcome, @Aaishpra . Look forward to your other contribution to the project.

File: src/tests/Register.test.js
Comment (mtreacy002) [inline src/tests/Register.test.js]: @Aaishpra , I just realized that these assert lines (143-147) need to be wrapped with the `await wait(()` to clear the warrning message (just like what you've done in PR #95 . Can you please fix this?

File: src/tests/Register.test.js
Comment (mtreacy002) [inline src/tests/Register.test.js]: Same here, please wrap in `await wait(()` to remove warning.

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
@Aaishpra . I've just realized one thing when reviewing your PR #95 . Please check my comment below.

File: src/tests/Register.test.js
Comment (Aaishpra) [reply src/tests/Register.test.js]: Sure

**Comment by Aaishpra** [comment]
Done @mtreacy002

File: src/tests/Register.test.js
Comment (mtreacy002) [reply src/tests/Register.test.js]: Great, @Aaishpra , just one minor thing. Indent the assert lines, please (here as well as on lines 186-190). Once you done them, I'll approve asap. 😉

**Review by mtreacy002 — CHANGES_REQUESTED** [review]
Hi @Aaishpra , thanks for doing the change request swiftly. Just one very minor thing left. See my comment below.

File: src/tests/Register.test.js
Comment (Aaishpra) [reply src/tests/Register.test.js]: > Great, @Aaishpra , just one minor thing. Indent the assert lines, please (here as well as on lines 186-190). Once you done them, I'll approve asap. wink

Added the changes.

**Review by mtreacy002 — DISMISSED** [review]
@Aaishpra . Thank you for your swift response. I'm really happy and glad to have you as part of our Open Source community.

**Comment by Aaishpra** [comment]
@mtreacy002 @Kajol-Kumari Please have a look here also.

**Review by mtreacy002 — DISMISSED** [review]
@Aaishpra, thanks for fixing the line. This LGTM now.

**Comment by mtreacy002** [comment]
@Aaishpra , can you please resolve the merge conflict? This conflict happens because there's another PR that just got merged to the develop branch that happened to have changes in the same file you are working on. This is a common thing in a team / open source environment. You just need to resolve it by taking into account the changes that recently got merged to your code base. 
Let me know if you are not sure how to do this with git pull --rebase. Also, I suggest you squash all your commit into one commit before using git pull --rebase from the upstream develop branch. Because otherwise you'll face a 'rebase hell` depending on the number of commits you have on your PR branch (in your case this PR has 22 commits, meaning that you have to do solve resolve conflict 22 times which would be crazy). 
Remember, call out if you need help. I'm more than happy to guide you through the process. 😉

**Comment by Aaishpra** [comment]
Hey, @mtreacy002 I resolved the conflicts directly through Github since I was getting an error while checking out the branch. Can you help in squashing these commits as I am having trouble while checking out my branch in the terminal?

**Comment by mtreacy002** [comment]
> Hey, @mtreacy002 I resolved the conflicts directly through Github since I was getting an error while checking out the branch. Can you help in squashing these commits as I am having trouble while checking out my branch in the terminal?

Sure, @Aaishpra . No worries. I'll squash and merge after I review the change 👍.

**Comment by Aaishpra** [comment]
Thanks for your help @mtreacy002

File: src/register/Register.jsx
Comment (Kajol-Kumari) [inline src/register/Register.jsx]: can we please use camelCase here (line 165 and 166)? `confirmPassword` instead of `confirmpassword`.

**Review by Kajol-Kumari — CHANGES_REQUESTED** [review]
@Aaishpra thank you for your hard work, Just a minor change is required, rest looks good me :)

File: src/register/Register.jsx
Comment (Aaishpra) [reply src/register/Register.jsx]: Sure

Comment (Aaishpra) [reply src/register/Register.jsx]: > can we please use camelCase here (line 165 and 166)? `confirmPassword` instead of `confirmpassword`.

Done, Please review

**Comment by Aaishpra** [comment]
@Kajol-Kumari please have a look.

File: src/register/Register.jsx
Comment (Kajol-Kumari) [inline src/register/Register.jsx]: Can we please fix this typo here? `formUserName` from `formUserame` if this change don't require any change in other files.

Comment (Aaishpra) [reply src/register/Register.jsx]: > Can we please fix this typo here? `formUserName` from `formUserame` if this change don't require any change in other files.

Fixed that, in my view, since the build is not failing it didn't require change in any other file.

**Comment by Aaishpra** [comment]
@Kajol-Kumari have a look

**Review by Kajol-Kumari — APPROVED** [review]
Thanks you for incorporating all the changes. LGTM now :)

**Comment by mtreacy002** [comment]
> Thanks you for incorporating all the changes. LGTM now :)

@Kajol-Kumari, can you please tick the approve box for the review post, otherwise it won't count as approval.

**Comment by Kajol-Kumari** [comment]
> > Thanks you for incorporating all the changes. LGTM now :)
> 
> @Kajol-Kumari, can you please tick the approve box for the review post, otherwise it won't count as approval.

Hey @mtreacy002 I have checked the approval box but as I don't have the write access to this repo as you do so it's not appearing as a green tick and instead appearing as a grey tick.

**Review by mtreacy002 — APPROVED** [review]
LGTM too, @Aaishpra . Thanks again for your contribution 👍

**Comment by mtreacy002** [comment]
I see. Thanks, @Kajol-Kumari for your explanation. This confirms my suspicion that only reviews given by maintainers/admins (who no doubt have write access to the repo) counted towards number pf appr

> > > Thanks you for incorporating all the changes. LGTM now :)
> > 
> > 
> > @Kajol-Kumari, can you please tick the approve box for the review post, otherwise it won't count as approval.
> 
> Hey @mtreacy002 I have checked the approval box but as I don't have the write access to this repo as you do so it's not appearing as a green tick and instead appearing as a grey tick.

I see. Thanks for your explanation, @Kajol-Kumari. This confirms my suspicion that only reviews from maintainers and admins are counted towards the numbers of approval. @isabelcosta, @mayburgos, and @annabauza, is there any way to solve this issue? Can you please fix this by giving `write` permission to the OS teams as well, so that their review counted towards the number of approval? Otherwise this can be a blocker for the project.

* Approval from admin counted towards number of approval.
Screen Shot 2020-09-17 at 6 56 37 pm

* Approval from @anitab-org/coding-team atm doesn't count towards the number of approval.
Screen Shot 2020-09-17 at 6 57 06 pm

Screen Shot 2020-09-17 at 7 01 33 pm

**Comment by mtreacy002** [comment]
@Kajol-Kumari , can you please try to re approve? since @annabauza just changed the approval [here](https://anitab-org.zulipchat.com/#narrow/stream/216321-coding/topic/Asking.20for.20advice/near/210366003), so maybe your previous approval won't be taken into account since it's made before the access permission change.

**Review by Kajol-Kumari — APPROVED** [review]

**Comment by Kajol-Kumari** [comment]
@mtreacy002 It will only show up if given write access :)

**Comment by mtreacy002** [comment]
> LGTM 👍

Thanks, @Kajol-Kumari . @annabauza It still doesn't work. The approval still show as `grey` aka `doesn't get counted towards number of approval`

**Review by annabauza — APPROVED** [review]

**Comment by mtreacy002** [comment]
Thanks, @annabauza for helping with the approval 👍

**Pull request merged**
